### ISSUE TITLE
derivable_impls reports Default when implementing in test submodule

### ISSUE THREAD

**Issue by Javex** [rust-lang/rust-clippy#12546]
### Summary

I have a `struct Foo` in my module which has quite a few fields. To ease testing, I implemented `Default` within `mod test`. This is not functionality I want to expose outside of tests, but it is a nice convenience when building objects for testing. However, clippy notes I could derive the implementation, not taking into account that I don't want to expose this functionality outside of tests.

### Lint Name

derivable_impls

### Reproducer

I tried this code:

```rust
struct Foo {
    bar: bool,
}

fn main() {}

#[cfg(test)]
mod test {
    use super::*;

    impl Default for Foo {
        fn default() -> Self {
            Self { bar: false }
        }
    }
    #[test]
    fn test() {
        let f = Foo::default();
        assert!(!f.bar);
    }
}
```

I saw this happen:

```
warning: this `impl` can be derived
  --> src/main.rs:11:5
   |
11 | /     impl Default for Foo {
12 | |         fn default() -> Self {
13 | |             Self { bar: false }
14 | |         }
15 | |     }
   | |_____^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#derivable_impls
   = note: `#[warn(clippy::derivable_impls)]` on by default
   = help: remove the manual implementation...
help: ...and instead derive it
   |
1  + #[derive(Default)]
2  | struct Foo {
   |

warning: struct `Foo` is never constructed
 --> src/main.rs:1:8
  |
1 | struct Foo {
  |        ^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: `clippy-repro` (bin "clippy-repro") generated 1 warning
warning: `clippy-repro` (bin "clippy-repro" test) generated 1 warning (run `cargo clippy --fix --bin "clippy-repro" --tests` to apply 1 suggestion)
    Finished dev [unoptimized + debuginfo] target(s) in 0.07s

```

I expected to see this happen: No lint warning for the default since it is implemented in sub module (or specifically, a `cfg(test)` module).

### Version

```text
rustc 1.77.0 (aedd173a2 2024-03-17)
binary: rustc
commit-hash: aedd173a2c086e558c2b66d3743b344f977621a7
commit-date: 2024-03-17
host: x86_64-unknown-linux-gnu
release: 1.77.0
LLVM version: 17.0.6
```

**Comment by Centri3** [comment]
Can't you do this with `#[cfg_attr(test, derive(Default))]`? Seems more like a bad suggestion instead of a FP

**Comment by Centri3** [comment]
Not sure how we'd make a proper suggestion for this tho, since these can be nested ig. e.g., `cfg(windows)` on a module, then `cfg(test)`.

**Comment by Javex** [comment]
Oh I didn't know that was possible (still new to rust)! Yes that works like a charm. In terms of suggestion, could it just be documented on https://rust-lang.github.io/rust-clippy/master/index.html#/derivable_impls? Not sure if this is too specific, but if that page had suggested using conditional compilation that would've been enough to point me in the right direction.

**Comment by Centri3** [comment]
Yeah I was thinking we could if we can't make an accurate suggestion (or even help message). But also it's an edge case and we tend not t o put those there.
